### ISSUE TITLE
Function to convert IPv6 to appropriate IPv4/6 string?

### ISSUE THREAD

**Issue by mzealey** [ClickHouse/ClickHouse#20469]
Following on from #19518, it would be nice to have a reverse function which operates on an ipv6 address and returns it in ipv4 or ipv6 notation depending on type. So basically `replaceOne(IPv6NumToString(ip), '::ffff:', '')`. This should then allow the `IPv6` type to work pretty much the same as the postgres `inet` type I think?

**Comment by bharatnc** [comment]
Yes we will need `toIPv4`,  `FunctionIPv4StringToNum`  mirror the current behavior  `toIPv6` & `FunctionIPv6StringToNum`. 

Currently toIPv4 does:

```sql
SELECT toIPv4('::ffff:192.168.1.1')
┌─toIPv4('::ffff:192.168.1.1')─┐
│                      0.0.0.0 │
└──────────────────────────────┘
```

Instead it should do:

```sql
SELECT toIPv4('::ffff:192.168.1.1')

┌─toIPv4('::ffff:192.168.1.1')─┐
│                      192.168.1.1 │
└──────────────────────────────┘
```

I can add this if someone can /assign this to me (related to my previous work on #19518).

**Comment by mzealey** [comment]
Maybe, however my request was for a function that effectively doesn't care if the source is ipv4 or ipv6 (within an ipv6 domain) but returns an ipv4 string if it is ipv4 address. That way and with the new toipv6 improvements in becomes trivial to dual stack stuff (with "legacy" clients expecting an ipv4 string) just using the ipv6 domain.On 14 Feb 2021 21:23, Bharat Nallan <notifications@github.com> wrote:

Yes we will need toIPv4,  FunctionIPv4StringToNum  mirror the current behavior  toIPv6 & FunctionIPv6StringToNum.

Currently toIPv4 does:

SELECT toIPv4('::ffff:192.168.1.1')

┌─toIPv4('::ffff:192.168.1.1')─┐

│                      0.0.0.0 │

└──────────────────────────────┘

Instead it should do:

SELECT toIPv4('::ffff:192.168.1.1')



┌─toIPv4('::ffff:192.168.1.1')─┐

│                      192.168.1.1 │

└──────────────────────────────┘

I can add this if someone can /assign this to me (related to my previous work on #19518).

**Comment by filimonov** [comment]
So we need a function to convert IPv6 to string, which will drop `::ffff:` prefix for IPv4 addressed stored in IPv4. 

@mzealey any ideas how to name it?

**Comment by mzealey** [comment]
I'm not good with names - perhaps `IPv6NumToIPString` ?

**Comment by a-dot** [comment]
Any movement on this issue?

**Comment by CherkashinSergey** [comment]
Still no progress? We'd like to use this function.

**Comment by yakov-olkhovskiy** [comment]
Apparently this issue was abandoned. I think we can add ability for functions 'toIPv4()' to accept IPv6 type and proper ip4 string address in ip6 domain...

**Comment by yakov-olkhovskiy** [comment]
I decided to only allow cast from IPv6 type - casting from string would slowdown IPv4 parsing. So, in order to convert from IPv6 string in proper mapping block you will need to cast it to IPv6 first:
```
ip-172-31-42-195.us-east-2.compute.internal :) select toIPv4(toIPv6('::ffff:1.2.3.4'))

SELECT toIPv4(toIPv6('::ffff:1.2.3.4'))

Query id: 9b456d42-72f0-4ad0-a68a-d3feb266d0bb

┌─toIPv4(toIPv6('::ffff:1.2.3.4'))─┐
│ 1.2.3.4                          │
└──────────────────────────────────┘

1 row in set. Elapsed: 0.006 sec.
```

**Comment by mzealey** [comment]
The original request was for a function which basically 'formats ipv6 as ipv6, or ::ffff:... in ipv4 notation'. Not really anything to do with cross-conversion of the backend data types, useful as that would be.

**Comment by yakov-olkhovskiy** [comment]
Well, it's easily achievable with UDF:
```
CREATE FUNCTION toIPv4OrIPv6String AS ip6str -> if(startsWith(ip6str, '::ffff:'), substr(ip6str, 8), ip6str);

WITH toString(toIPv6('::ffff:0102:0304')) AS ip6 SELECT toIPv4OrIPv6String(ip6)

┌─if(startsWith(ip6, '::ffff:'), substring(ip6, 8), ip6)─┐
│ 1.2.3.4                                                │
└────────────────────────────────────────────────────────┘
```

**Comment by mzealey** [comment]
Sure, or the suggestion I had in my initial report. But if you were able to implement it as a core function (perhaps via UDF route you describe above) it would make it a lot simpler to port postgres data over as it could nicely emulate the `inet` type functionality in an IPv6 type

**Comment by CherkashinSergey** [comment]
I think, it would be nice to have some func like `IPv6ToIPv4()` that accepts an IPv6 num and converts it to IPv4 domain if provided IP is IPv4-mapped-to-IPv6. Maybe `isIPv4MappedToIPv6()` function would be useful too.

**Comment by yakov-olkhovskiy** [comment]
@CherkashinSergey it's exactly what https://github.com/ClickHouse/ClickHouse/pull/49759 meant to do
@mzealey but what's wrong with UDF? you create it once and then just use it...

**Comment by RunningFaster** [comment]
> #49759 I decided to only allow cast from IPv6 type - casting from string would slowdown IPv4 parsing. So, in order to convert from IPv6 string in proper mapping block you will need to cast it to IPv6 first:
> 
> ```
> ip-172-31-42-195.us-east-2.compute.internal :) select toIPv4(toIPv6('::ffff:1.2.3.4'))
> 
> SELECT toIPv4(toIPv6('::ffff:1.2.3.4'))
> 
> Query id: 9b456d42-72f0-4ad0-a68a-d3feb266d0bb
> 
> ┌─toIPv4(toIPv6('::ffff:1.2.3.4'))─┐
> │ 1.2.3.4                          │
> └──────────────────────────────────┘
> 
> 1 row in set. Elapsed: 0.006 sec.
> ```

What Version?

**Comment by yakov-olkhovskiy** [comment]
@RunningFaster this improvement is included since 23.6 - you can easily asses which version has feature you are looking for in the ClickHouse fiddle:
https://fiddle.clickhouse.com/27e41d90-af6b-4869-93f5-4b60e3271529
